crates/extensions/tedge_p11_server/src/main.rs

```diff
@@ -0,0 +1,163 @@
+//! thin-edge.io PKCS#11 server.
+//!
+//! The purpose of this crate is to allow thin-edge services that possibly run in containers to access PKCS#11 tokens in
+//! all of our supported architectures.
+//!
+//! There are 2 main problems with using a PKCS#11 module directly by thin-edge:
+//! 1. One needs to use a dynamic loader to load the PKCS#11 module, which is not possible in statically compiled musl
+//! 2. When thin-edge runs in a container, additional setup needs to be done by the user to expose cryptographic tokens
+//!    in the container, using software like p11-kit.
+//!
+//! To avoid extra dependencies and possibly implement new features in the future, it was decided that thin-edge.io will
+//! provide its own bundled p11-kit-like service.
+
+use std::io::{BufRead, BufReader, Read, Write};
+use std::os::unix::net::{UnixListener, UnixStream};
+use std::path::Path;
+use std::sync::Arc;
+
+use anyhow::Context;
+use camino::Utf8PathBuf;
+use certificate::parse_root_certificate::pkcs11::PkcsSigner;
+use certificate::parse_root_certificate::{pkcs11, CryptokiConfigDirect};
+use clap::command;
+use clap::Parser;
+use tracing::debug;
+use tracing::info;
+use tracing::level_filters::LevelFilter;
+use tracing_subscriber::EnvFilter;
+
+/// thin-edge.io service for passing PKCS#11 cryptographic tokens.
+#[derive(Debug, Clone, PartialEq, Eq, Parser)]
+#[command(version)]
+pub struct Args {
+    /// A path where the service's unix token will be created.
+    #[arg(default_value = "./thin-edge-pkcs11.sock")]
+    socket_path: Utf8PathBuf,
+
+    /// The path to the PKCS#11 module.
+    ///
+    /// If not provided, the module path will be read from tedge-config.
+    #[arg(long)]
+    module_path: Option<Utf8PathBuf>,
+
+    /// The PIN for the PKCS#11 token.
+    ///
+    /// If not provided, the PIN will be read from tedge-config.
+    #[arg(long, default_value = "123456")]
+    pin: Arc<str>,
+}
+
+fn main() -> anyhow::Result<()> {
+    tracing_subscriber::fmt()
+        .with_env_filter(
+            EnvFilter::builder()
+                .with_default_directive(LevelFilter::INFO.into())
+                .from_env()
+                .unwrap(),
+        )
+        .init();
+
+    let args = Args::parse();
+    let socket_path = args.socket_path;
+
+    if Path::new(&socket_path).exists() {
+        let _ = std::fs::remove_file(&socket_path);
+    }
+
+    let cryptoki_config = if let Some(module_path) = args.module_path {
+        CryptokiConfigDirect {
+            module_path,
+            pin: args.pin,
+            serial: None,
+        }
+    } else if let Some(cryptoki_config) = tedge_config::TEdgeConfigLocation::default()
+        .load()
+        .ok()
+        .and_then(|tedge_config| tedge_config.device.cryptoki.config_direct().ok().flatten())
```

Comment:
If we are reusing the tedge.toml, then we need to support the `--config-dir` argument which allows users to control where the base directory is, e.g. `/etc/tedge`...

Though I'd be open to discussing whether we actually need to read tedge.toml config at all.

Reply:
Further in the development depending on `tedge_config` in `tedge-p11-server` introduced a dependency cycle, and reading from `tedge-config` indeed does not provide much value (as this server runs on the host, separate from the rest of thin-edge services), so this functionality was removed.